free_fleet/src/manager/Manager.cpp

```diff
@@ -0,0 +1,435 @@
+void Manager::Implementation::handle_robot_state(
+  const messages::RobotState& state)
+{
+  std::lock_guard<std::mutex> lock(mutex);
+
+  const std::string robot_name = state.name();
+  const auto task_id = state.task_id();
+
+  messages::RobotState transformed_state(
+    state.time(),
+    robot_name,
+    state.model(),
+    task_id,
+    state.mode(),
+    state.battery_percent(),
+    to_robot_transform->backward_transform(state.location()),
+    state.target_path_index());
+
+  if (robot_name.empty())
+  {
+    fferr << "Incoming robot state's name is empty, ignoring.\n";
+    return;
+  }
+
+  const auto r_it = robots.find(robot_name);
+  bool new_robot = r_it == robots.end();
+  if (new_robot)
+  {
+    auto new_robot = manager::RobotInfo::Implementation::make(
+      transformed_state,
+      graph,
+      time_now_fn());
+    if (new_robot)
+    {
+      robots[robot_name] = std::move(new_robot);
+      ffinfo << "Registered new robot: [" << robot_name << "]\n";
+    }
+  }
+  else
+  {
+    manager::RobotInfo::Implementation::update_state(
+      *r_it->second,
+      transformed_state,
+      time_now_fn());
+  }
+
+  // Updates external uses of the robot's information
+  if (r_it != robots.end() && robot_updated_callback_fn)
+    robot_updated_callback_fn(*r_it->second);
+
+  // for each robot figure out whether any tasks were not received yet
+  if (task_id.has_value())
+  {
+    const auto t_it = unacknowledged_tasks.find(task_id.value());
+    if (t_it != unacknowledged_tasks.end())
+    {
+      t_it->second->acknowledge_request();
+      unacknowledged_tasks.erase(t_it);
+    }
+  }
+}
+
+//==============================================================================
+auto Manager::make(
+  const std::string& fleet_name,
+  std::shared_ptr<const rmf_traffic::agv::Graph> graph,
+  std::unique_ptr<transport::ServerMiddleware> middleware,
+  std::shared_ptr<const manager::CoordinateTransformer> to_robot_transform,
+  TimeNow time_now_fn,
+  RobotUpdatedCallback robot_updated_callback_fn)
+-> std::shared_ptr<Manager>
+{
+  auto make_error_fn = [](const std::string& error_msg)
+    {
+      fferr << error_msg << "\n";
+      return nullptr;
+    };
+
+  if (fleet_name.empty())
+    return make_error_fn("Provided fleet name must not be empty.");
+  if (!graph)
+    return make_error_fn("Provided traffic graph cannot be null.");
+  if (!middleware)
+    return make_error_fn("Provided free fleet middleware cannot be null.");
+  if (!to_robot_transform)
+    return make_error_fn(
+      "Provided free fleet manager::CoordinateTransformer cannot be null.");
+  if (!time_now_fn)
+    return make_error_fn("Provided time function cannot be null.");
+
+  std::shared_ptr<Manager> manager_ptr(new Manager);
+  manager_ptr->_pimpl->fleet_name = fleet_name;
+  manager_ptr->_pimpl->graph = std::move(graph);
+  manager_ptr->_pimpl->middleware = std::move(middleware);
```

Comment:
Down below I make the recommendation that `Implementation::middleware` should be managed by a `std::shared_ptr`. You can make that change while still having the `middleware` argument of this function be managed by a `std::unique_ptr`. All you need to do is change this line to:

```
manager_ptr->_pimpl->middleware = std::shared_ptr<transport::ServerMiddleware>(middleware.release());
```